File: leetcode/py/networkDelay.py
class Solution:
    def DFS(self, K, delay, graph):
        #if visited_map[K] == 1: return 0                
        for v, w in graph[K].items(): 
            if delay[K] + w < delay[v]:
                delay[v] = delay[K] + w
                self.DFS(v, delay, graph)#, visited_map)
        return -1
    
    def dijkstra(self, K, N, delay, graph, TIMEOUT):
        Q = set(range(N))                
        delay[K-1] = 0
        while len(Q) > 0:
            u = None
            for node in Q:
                if u == None or delay[node] < delay[u]:
                    u = node
            Q.remove(u)
            for v, w in graph[u].items():
                if delay[u] + w < delay[v]:
                    delay[v] = delay[u] + w
            #print(delay)
        d = max(delay)
        if d == TIMEOUT:
            d = -1
        return d
    
    def bellman_ford(self, K, N, delay, pre, graph, TIMEOUT):

        delay[K-1] = 0
        for i in range(N):
            for u, vs in graph.items():
                for v, w in vs.items():
                    if delay[v] > delay[u] + w:
                        delay[v] = delay[u] + w
            #print(delay)
        #for u, vs in graph.items():
        #    for v, w in vs:
        d = max(delay)
        if d == TIMEOUT:
            d = -1
        return d
    
    def networkDelayTime(self, times, N, K):
        """
        :type times: List[List[int]]
        :type N: int
        :type K: int
        :rtype: int
        """
        graph = {}
        delay = []
        pre = []
        TIME_OUT = 100000
        for it in range(N):
            graph[it] = {}           
            delay.append(TIME_OUT)
            pre.append(-1)
        for list_it in times:        
            u, v, w = list_it
            graph[u-1][v-1] = w
        #return self.dijkstra(K, N, delay, graph, TIME_OUT)#, visited_map)                
        return self.bellman_ford(K, N, delay, pre, graph, TIME_OUT)

File: leetcode/cpp/networkDelay.cc
using pairType = std::pair<int, int>;
//using std::unordered_map<int, std::unordered_map<int, int> > = graphType;
using graphType = std::unordered_map<int, vector<pairType> >; // std::unordered_map<int, int> > graphType;
using heapType = std::priority_queue<pairType, vector<pairType> >;

class Solution {

private:
    
    int dijkstra(const int & K, const int &N, std::vector<int> &delay, graphType & graph, const int &TIMEOUT ){
        int d = -1;
        heapType Q;
        delay[K]=0;
        Q.push(std::make_pair(0, K));
        while(Q.size()>0){
            const std::pair<int, int> dnode = Q.top();
            int node = dnode.second;
            Q.pop();
            const vector<pairType> vec_link = graph[node];
            for(const auto & vw : vec_link){
                    int v = vw.first;
                    int w = vw.second;
                    if(delay[node] + w < delay[v]){
                        delay[v] = delay[node] + w;
                        Q.push(std::make_pair(-w, v));
                    }
                } 
            /*
            cout<<"size "<<Q.size()<<endl;
            for(auto it : delay){
                cout<<it<<endl;
            }
            */
            
        }
        for(const auto & it : delay){
            if( it > d ) d = it;
            if( it == TIMEOUT) return -1; //delay = -1;
        }
        return d;
    }
    
public:
    int networkDelayTime(vector<vector<int>>& times, int N, int K) {
        const int TIMEOUT = 10000;
        std::vector<int> delay(N, TIMEOUT);
        graphType graph;
        
        for(int i=0; i < times.size(); i++){ 
            int u = times[i][0] -1;
            int v = times[i][1] -1;
            int w = times[i][2];
            graph[u].push_back(make_pair(v, w));
        }
        
        return dijkstra(K-1, N, delay, graph, TIMEOUT); 
    }
};
